@here/harp-datasource-protocol/lib/ExprEvaluator.ts

```diff
@@ -69,64 +77,406 @@ export class ExprEvaluator implements ExprVisitor<Value, Env> {
+                    if (!this.evaluate(childExpr, env)) {
+                        return false;
+                    }
                 }
-                return false;
-            }
-            case "^=": {
-                if (typeof left === "string" && typeof right === "string") {
-                    return left.startsWith(right);
+                return true;
+
+            case "any":
+                for (const childExpr of expr.children) {
+                    if (this.evaluate(childExpr, env)) {
+                        return true;
+                    }
                 }
                 return false;
-            }
-            case "$=": {
-                if (typeof left === "string" && typeof right === "string") {
-                    return left.endsWith(right);
+
+            case "none":
+                for (const childExpr of expr.children) {
+                    if (this.evaluate(childExpr, env)) {
+                        return false;
+                    }
                 }
-                return false;
+                return true;
+
+            case "boolean":
+            case "number":
+            case "string":
+                for (const childExpr of expr.children) {
+                    const value = this.evaluate(childExpr, env);
+                    if (typeof value === expr.op) {
+                        return value;
+                    }
+                }
+                throw new Error(`expected a '${expr.op}'`);
+
+            default: {
+                const descriptor = builtinFunctions.get(expr.op);
+                if (descriptor) {
+                    const actuals = expr.children.map(arg => this.evaluate(arg, env));
+                    return descriptor.call(actuals);
+                }
+                throw new Error(`undefined operator '${expr.op}`);
             }
-            case "==":
-                return left === right;
-            case "!=":
-                return left !== right;
-            case "<":
-                return left !== undefined && right !== undefined ? left < right : undefined;
-            case ">":
-                return left !== undefined && right !== undefined ? left > right : undefined;
-            case "<=":
-                return left !== undefined && right !== undefined ? left <= right : undefined;
-            case ">=":
-                return left !== undefined && right !== undefined ? left >= right : undefined;
-        }
-        throw new Error(`invalid relational op '${expr.op}'`);
-    }
-
-    visitLogicalExpr(expr: LogicalExpr, env: Env): Value {
-        const value = this.evaluate(expr.left, env);
-        switch (expr.op) {
-            case "||":
-                return value || this.evaluate(expr.right, env);
-            case "&&":
-                return value && this.evaluate(expr.right, env);
         } // switch
-        throw new Error(`invalid logical op '${expr.op}'`);
     }
 }
+
+ExprEvaluator.registerBuiltin("length", {
```

Comment:
Maybe we should store all this builtin tokens in some enum, and address them like ```ExprBuiltin.Length``` for example?

Reply:
The code was reorganized a bit, now separate files expose the builtin operators and a typed string enumeration (e.g. `CastOperators.ts` exports the `CastOperators` builtins and the `CastOperatorNames` enumeration).